spec/src/main/asciidoc/query-language.asciidoc

```diff
@@ -0,0 +1,337 @@
+- operator expressions.
+
+==== Literal expressions
+
+A string, integer, or decimal literal is assigned the type it would be assigned in Java. So, for example, `'Hello'` is assigned the type `java.lang.String`, `123` is assigned the type `int`, `1e4` is assigned the type `double`, and `1.23f` is assigned the type `float`.
+
+The syntax for literal expressions is given by the `literal` grammar rule, and in the previous section titled <<Lexical structure>>.
+
+When executed, a literal expression evaluates to its literal value.
+
+==== Special values
+
+The special values `true` and `false` are assigned the type `boolean`, and evaluate to their literal values.
+
+The special values `local date` and `local datetime` are assigned the types `java.time.LocalDate` and `java.time.LocalDateTime`, and evaluate to the current date and current datetime of the database server, respectively.
+
+The syntax for special values is given by the `special_expression` grammar rule.
+
+==== Parameter expressions
+
+A parameter expression, with syntax given by `input_parameter`, is assigned the type of the repository method parameter it matches. For example, the parameter `:titlePattern` is assigned the type `java.lang.String`:
+
+[source,java]
+----
+@Query("where title like :titlePattern")
+List<Book> booksMatchingTitle(String titlePattern);
+----
+
+When executed, a parameter expression evaluates to the argument supplied to the parameter of the repository method.
+
+==== Enum literals
+
+An _enum literal expression_ is a Java identifier, with syntax specified by `enum_literal`, and may only occur as the right operand of a `set` assignment or `=`/`<>` equality comparison. It is assigned the type of the left operand of the assignment or comparison. The type must be a Java `enum` type, and the identifier must be the name of an enumerated value of the `enum` type. For example, `day <> MONDAY` is a legal comparison expression.
+
+When executed, an enum expression evaluates to the named member of the Java `enum` type.
+
+==== Path expressions
+
+A _path expression_ is a period-separated list of Java identifiers, with syntax specified by `state_field_path_expression`. Each identifier is interpreted as the name of a field of an entity or embeddable class. Each prefix of the list is assigned a Java type:
+
+- the first element of the list is assigned the type of the named field of the entity being queried, and
+- each subsequent element is assigned the type of the named field of the type assigned to the previous element.
+
+The type of the whole path expression is the type of the last element of the list. For example, `pages` is assigned the type `int`, `address` is assigned the type `org.example.Address`, and `address.street` is assigned the type `java.lang.String`.
+
+When executed, a path expression is evaluated in the context of a given record of the queried entity type, and evaluates to the value of the entity field for the given record.
+
+==== Function calls
```

Comment:
About those functions, there is no guarantee of this function support on the NoSQL database side.

Reply:
Yeah, so what I want to know is which functions _can_ be implemented, "typically"? Or at least, if not "typically", which can be implemented for a sufficient subset of platforms that it's worth mentioning them here/

Reply:
I mean, if the answer is simply that none of these things can typically be implemented on non-relational databases, then we should simply remove this section.

Reply:
Typically at Document and Graph:

| Function | Supported by | Not Supported by |
|----------|--------------|------------------|
| `abs()`  | Document, Graph | Wide-column, Key-value |
| `length()`| Document, Graph | Wide-column, Key-value |
| `lower()` | Document, Graph | Wide-column, Key-value |
| `upper()` | Document, Graph | Wide-column, Key-value |
| `left()`  | Document, Graph | Wide-column, Key-value |
| `right()` | Document, Graph | Wide-column, Key-value |

But even with document, I cannot guarantee, e.g.:

- ravendb
- harperdb

Reply:
All functions, and an UnsupportedOperationException is thrown only when the database provider does not support it.

Reply:
All the functions on that list?

Reply:
Yes!

- abs
- length
- lower
- upper
- left
- right

Reply:
Can `local date` and `local datetime` be supported on all platforms? (I'm guessing "no", but looking for confirmation.)

Reply:
Please see commit https://github.com/jakartaee/data/pull/520/commits/858686efec3b0f261d464851156e3d58fad4037c.

Reply:
can we mandate compilation time checking of supported functions? runtime errors are not nice

Reply:
Well the issue is that at the spec level, Jakarta Data allows purely-runtime implementations.

Now, as far as I can tell, we're all implementing the spec using annotation processing, and so, indeed, we can all check this at compilation time (I certainly do).

But if we were to _require_ that, then we would be prohibiting other implementation strategies.

Reply:
As discussed on a recent concall, I already added `local time`. And it seems we have a better-than-tentative consensus on the proposed list of functions.